### PR TITLE
[SPARK-7007][core] Add a metric source for ExecutorAllocationManager

### PR DESCRIPTION
**Pull request by jerryshao**
Add a metric source to expose the internal status of ExecutorAllocationManager to better monitoring the resource usage of executors when dynamic allocation is enable. Please help to review, thanks a lot.

### CODE REVIEW

**Comment by jerryshao** [comment]
Jenkins, retest this please.

**Comment by andrewor14** [comment]
@jerryshao this looks good. It will allow us to plot these variables in very interesting ways. Note that this relies on internal variables that are subject to change. For instance, #5536 gets rid of `numExecutorsPending`. Maybe we should add a comment at the top to indicate that this is unstable across Spark versions.

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala [hunk 1]
```diff
@@ -135,6 +138,26 @@ private[spark] class ExecutorAllocationManager(
   private val executor = Executors.newSingleThreadScheduledExecutor(
     Utils.namedThreadFactory("spark-dynamic-executor-allocation"))
 
+  // Metric source for ExecutorAllocationManager to expose the its internal executor allocation
+  // status to MetricsSystem.
+  private[spark] val executorAllocationManagerSource = new Source {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: Would you mind putting this as an inner class at the bottom (like `ExecutorAllocationManagerListener`) instead?

**Comment by andrewor14** [comment]
Also it seems that you need to rebase this one.

**Comment by jerryshao** [comment]
Hi @andrewor14 , thanks a lot for your comments, I will change the code accordingly.

**Comment by jerryshao** [comment]
Hi @andrewor14 , I've changed the code according to your comments, please take a look at this, thanks a lot.

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala [hunk 2]
```diff
@@ -135,6 +138,9 @@ private[spark] class ExecutorAllocationManager(
   private val executor =
     ThreadUtils.newDaemonSingleThreadScheduledExecutor("spark-dynamic-executor-allocation")
 
+  // Metric source for ExecutorAllocationManager to expose internal status to MetricsSystem.
+  private[spark] val executorAllocationManagerSource = new ExecutorAllocationManagerSource
```
Comment (sryza) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: The class is already `private[spark]`, so we don't need to declare this again here.

File: core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala [hunk 3]
```diff
@@ -582,6 +588,26 @@ private[spark] class ExecutorAllocationManager(
     }
   }
 
+  // Metric source for ExecutorAllocationManager to expose the its internal executor allocation
```
Comment (sryza) [inline core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala]: Nit: "the" and "its" are redundant.  Also, this should be a block-style comment.

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 4]
```diff
@@ -518,6 +518,9 @@ class SparkContext(config: SparkConf) extends Logging with ExecutorAllocationCli
     _taskScheduler.postStartHook()
     _env.metricsSystem.registerSource(new DAGSchedulerSource(dagScheduler))
     _env.metricsSystem.registerSource(new BlockManagerSource(_env.blockManager))
+    _executorAllocationManager.foreach { e =>
```
Comment (sryza) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: The model for the other metric sources seems to be that the source wraps the object.  What's the reason for having the source inside of the ExecutorAllocationManager in this case?

Comment (jerryshao) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: Hi @sryza , thanks a lot for your comments. The reason I put this metric source into `ExecutorAllocationManager` is that I don't want to change `ExecutorAllocationManager` a lot to expose the internal status to metric source, so I just wrote this metric source as the inner class. Do you have specific concern about this implementation?

**Comment by andrewor14** [comment]
@jerryshao I believe this doesn't compile anymore since I merged #5074. Would you mind rebasing this?

**Comment by sryza** [comment]
@jerryshao no specific concerns about the implementation.  If @andrewor14 is ok with the inconsistency, I'm fine with it too.

**Comment by andrewor14** [comment]
Yeah it's actually going to be a little tricky to bring this outside because this accesses a whole bunch of internal variables. I even think that we should add a big comment at the top of the source class to say this is experimental and there are no backward compatibility guarantees at all.

**Comment by jerryshao** [comment]
@andrewor14 @sryza , thanks a lot for your comments, I just rebase the codes, please review it.

**Comment by andrewor14** [comment]
Looks great. Merging into master and 1.4 thanks @sryza @jerryshao

**Pull request closed without merging**